executor/new_distsql.go

```diff
@@ -317,28 +317,33 @@ func (e *IndexLookUpExecutor) startIndexWorker(kvRanges []kv.KeyRange, workCh ch
 		return errors.Trace(err)
 	}
 	result.Fetch(e.ctx.GoCtx())
-	ih := &e.indexWorker
-	ih.wg.Add(1)
+	worker := &e.indexWorker
+	worker.wg.Add(1)
 	go func() {
 		ctx, cancel := goctx.WithCancel(e.ctx.GoCtx())
-		ih.fetchHandles(e, result, workCh, ctx, finished)
+		worker.fetchHandles(e, result, workCh, ctx, finished)
 		cancel()
 		if err := result.Close(); err != nil {
 			log.Error("close SelectDAG result failed:", errors.ErrorStack(err))
 		}
 		close(workCh)
-		ih.wg.Done()
+		close(e.resultCh)
+		worker.wg.Done()
 	}()
 	return nil
 }
 
 // fetchHandles fetches a batch of handles from index data and builds the index lookup tasks.
-func (ih *indexWorker) fetchHandles(e *IndexLookUpExecutor, result distsql.SelectResult, workCh chan<- *lookupTableTask, ctx goctx.Context, finished <-chan struct{}) {
+// The tasks are sent to workCh to be further processed by tableWorker, and sent to e.resultCh
+// at the same time to keep data ordered.
+func (worker *indexWorker) fetchHandles(e *IndexLookUpExecutor, result distsql.SelectResult, workCh chan<- *lookupTableTask, ctx goctx.Context, finished <-chan struct{}) {
 	for {
 		handles, finish, err := extractHandlesFromIndexResult(result)
 		if err != nil {
-			workCh <- &lookupTableTask{
-				tasksErr: errors.Trace(err),
+			doneCh := make(chan error, 1)
+			doneCh <- errors.Trace(err)
+			e.resultCh <- &lookupTableTask{
```

Comment:
Need to select `finished` channel.

Reply:
If `finished` is closed, means we've close the `IndexLookUpExecutor` and don't care the result any more.
`IndexLookUpExecutor.Close()` would drain the remain lookupTableTask and discard them.
Both the same result, so select `finished` here or not doesn't make any difference.

Reply:
ok